pkg/envoy/agent.go

```diff
@@ -105,13 +138,71 @@ func (a *Agent) terminate() {
 	if e != nil {
 		log.Warnf("Error in invoking drain listeners endpoint %v", e)
 	}
-	log.Infof("Graceful termination period is %v, starting...", a.terminationDrainDuration)
-	time.Sleep(a.terminationDrainDuration)
-	log.Infof("Graceful termination period complete, terminating remaining proxies.")
-	a.abortCh <- errAbort
+	log.Infof("Agent draining proxy for %v, then waiting for active connections to terminate...", a.minDrainDuration)
+	time.Sleep(a.minDrainDuration)
+	log.Infof("Termination drain duration period is %v, checking for active connections...", a.terminationDrainDuration)
+	go a.waitForDrain()
+	select {
+	case <-a.drainCh:
+		log.Info("There are no more active connections. terminating proxy...")
+		a.abortCh <- errAbort
+	// TODO: remove terminationDrainDuration and rely on "terminationGracefulPeriodSeconds" of pod?
+	case <-time.After(a.terminationDrainDuration):
+		log.Info("Termination period complete, terminating proxy...")
+		a.abortCh <- errAbort
+		close(a.drainCh)
+	}
 	log.Warnf("Aborted all epochs")
 }
 
+func (a *Agent) waitForDrain() {
+	for {
+		select {
+		case <-time.After(activeConnectionCheckDelay):
+			if a.activeProxyConnections() == 0 {
+				close(a.drainCh)
+				return
+			}
+		case <-a.drainCh:
+			return
+		}
+	}
+}
+
+func (a *Agent) activeProxyConnections() int {
```

Comment:
Is it possible users are excluding `downstream_cx_active` stats? I actually am not seeing it at all in a default install...

Reply:
yes. We will have to add them to default inclusion

Reply:
It seems like we haven't done that. And I am a bit concerned about adding `downstream_cx_active` for *all* listeners. Can we just add it for the 2 listeners we need it for?

Reply:
I have not added it yet. I will do that. Yes. We can add it for two listeners

Reply:
We don’t have that many inbound listeners so adding this metric is fine.
We were considering adding downstream connection metrics for a long time anyway.
It is especially useful at ingress.

Reply:
Can you check active vs open, and make sure it captures the correct quantity as the name implies.

Reply:
If you mean total vs active - I verified. It captures the correct quantity